repose-aggregator/components/filters/attribute-mapping-policy-validation-filter/src/main/scala/org/openrepose/filters/attributemapping/AttributeMappingPolicyValidationFilter.scala

```diff
@@ -0,0 +1,122 @@
+
+  import AttributeMappingPolicyValidationFilter._
+
+  // Create a transform factory using the default implementation and the classloader for this class
+  private val transformerFactory = TransformerFactory.newInstance(
+    "org.apache.xalan.xsltc.trax.TransformerFactoryImpl",
+    this.getClass.getClassLoader)
+
+  override def init(filterConfig: FilterConfig): Unit = {}
+
+  override def destroy(): Unit = {}
+
+  override def doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = {
+    val httpServletRequest = request.asInstanceOf[HttpServletRequest]
+    val httpServletResponse = response.asInstanceOf[HttpServletResponse]
+    val requestContentType = httpServletRequest.getContentType
+
+    // Buffer the request payload so that we can validate the policy, but forward the original payload
+    val requestPayloadBuffer = new BufferedServletInputStream(httpServletRequest.getInputStream)
+    requestPayloadBuffer.mark(Integer.MAX_VALUE)
+
+    validateHttpMethod(httpServletRequest) flatMap { _ =>
+      getPolicyAsXmlSource(requestContentType, httpServletRequest.getInputStream)
+    } map { policyXmlSource =>
+      AttributeMapper.validatePolicy(policyXmlSource, XSDEngine.AUTO.toString)
+    } match {
+      case Success(_) =>
+        requestPayloadBuffer.reset()
+        chain.doFilter(
+          new HttpServletRequestWrapper(
+            httpServletRequest,
+            requestPayloadBuffer),
+          response)
+      case Failure(exception) =>
+        exception match {
+          case uhme: UnsupportedHttpMethodException =>
+            logger.debug("Unsupported HTTP method -- no validation performed", uhme)
+            chain.doFilter(request, response)
+          case ucte: UnsupportedContentTypeException =>
+            logger.debug("Unsupported Content-Type -- validation failed", ucte)
+            httpServletResponse.sendError(
+              SC_UNSUPPORTED_MEDIA_TYPE,
+              ucte.message)
+          case e@(_: SaxonApiException | _: TransformerException | _: JsonProcessingException | _: JsonParseException) =>
+            logger.debug("Validation failed", e)
+            httpServletResponse.sendError(
+              SC_BAD_REQUEST,
+              "Failed to validate attribute mapping policy in request")
+        }
+    }
+  }
+
+  def validateHttpMethod(request: HttpServletRequest): Try[Unit.type] = {
+    if ("PUT".equalsIgnoreCase(request.getMethod)) Success(Unit)
+    else Failure(UnsupportedHttpMethodException(s"${request.getMethod} is not a supported HTTP method"))
+  }
+
+  def getPolicyAsXmlSource(contentType: String, policy: InputStream): Try[Source] = {
+    val contentTypeLowerCase = contentType.toLowerCase
+    val requestStreamSource = new StreamSource(policy)
+
+    if (contentTypeLowerCase.contains("json")) {
+      Try {
+        val outPolicyXml = new XdmDestination
+        AttributeMapper.policy2XML(requestStreamSource, outPolicyXml)
```

Comment:
My primary motivation was to save time. I wanted to get this story wrapped up by end-of-sprint, which is today. Pushing this logic into attribute mapper would require that I spend a bit of time figuring out more about that library (e.g., the testing framework). It would also require a release of that library.

Instead, I took the approach I would take with any third-party library -- I leveraged the functionality that currently exists. Of course, that does not prevent me (or us) from making a PR to that library in the future and adapting our code when we do. That is, I looked at this iteratively and chose to make a small first step, after which more steps could follow.